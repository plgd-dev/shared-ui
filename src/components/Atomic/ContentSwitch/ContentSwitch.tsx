import React, { FC, Suspense, useCallback } from 'react'
import { motion, AnimatePresence } from 'framer-motion'
import isFunction from 'lodash/isFunction'

import { Props } from './ContentSwitch.types'
import ConditionalWrapper from '../ConditionalWrapper/ConditionalWrapper'

const ContentSwitch: FC<Props> = (props) => {
    const { activeItem, children, isAsync, onAnimationComplete } = props

    const onAnimationCompleteCallback = useCallback(() => {
        isFunction(onAnimationComplete) && onAnimationComplete()
    }, [onAnimationComplete])

    return (
        <ConditionalWrapper condition={isAsync} wrapper={(c) => <Suspense fallback={<div />}>{c} </Suspense>}>
            <AnimatePresence initial={false} mode='wait' onExitComplete={() => null}>
                {React.Children.map(children, (child, i) => {
                    if (i !== activeItem) {
                        return null
                    }
                    return (
                        <motion.div
                            animate={{ opacity: 1 }}
                            exit={{ opacity: 0 }}
                            initial={{ opacity: 0 }}
                            onAnimationComplete={onAnimationCompleteCallback}
                            transition={{
                                duration: 0.3,
                            }}
                        >
                            {child}
                        </motion.div>
                    )
                })}
            </AnimatePresence>
        </ConditionalWrapper>
    )
}

ContentSwitch.displayName = 'ContentSwitch'

export default ContentSwitch
