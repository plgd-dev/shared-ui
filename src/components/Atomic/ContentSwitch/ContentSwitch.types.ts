import { CSSProperties, ReactElement } from 'react'

export type Props = {
    activeItem: number
    children: ReactElement[]
    isAsync?: boolean
    onAnimationComplete?: () => void
    style?: CSSProperties
}
