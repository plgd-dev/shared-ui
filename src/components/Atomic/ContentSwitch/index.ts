export { default } from './ContentSwitch'
export * from './ContentSwitch'
