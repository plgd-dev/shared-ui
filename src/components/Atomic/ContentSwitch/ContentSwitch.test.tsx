import { act, waitFor } from '@testing-library/react'

import { render } from '../../../../test/jest-wrapper'
import ContentSwitch from './ContentSwitch'

describe('<ContentSwitch>', () => {
    it('renders without crashing', () => {
        const { asFragment } = render(
            <ContentSwitch activeItem={0}>
                <div>Item 1</div>
                <div>Item 2</div>
            </ContentSwitch>
        )

        expect(asFragment()).toMatchSnapshot()
    })

    it('renders the active item', () => {
        const { getByText } = render(
            <ContentSwitch activeItem={0}>
                <div>Item 1</div>
                <div>Item 2</div>
            </ContentSwitch>
        )
        expect(getByText('Item 1')).toBeInTheDocument()
    })

    it('does not render inactive items', () => {
        const { queryByText } = render(
            <ContentSwitch activeItem={0}>
                <div>Item 1</div>
                <div>Item 2</div>
            </ContentSwitch>
        )
        expect(queryByText('Item 2')).toBeNull()
    })

    it('calls onAnimationComplete when animation completes', async () => {
        const onAnimationComplete = jest.fn()
        const { rerender } = render(
            <ContentSwitch activeItem={0} onAnimationComplete={onAnimationComplete}>
                <div>Item 1</div>
                <div>Item 2</div>
            </ContentSwitch>
        )

        rerender(
            <ContentSwitch activeItem={1} onAnimationComplete={onAnimationComplete}>
                <div>Item 1</div>
                <div>Item 2</div>
            </ContentSwitch>
        )

        act(() => {
            jest.runAllTimers()
        })

        await waitFor(() => {
            expect(onAnimationComplete).toHaveBeenCalled()
        })
    })
})
