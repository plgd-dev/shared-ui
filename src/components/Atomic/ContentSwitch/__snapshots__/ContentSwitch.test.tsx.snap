// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentSwitch> renders without crashing 1`] = `
<DocumentFragment>
  <div
    style="opacity: 1;"
  >
    <div>
      Item 1
    </div>
  </div>
</DocumentFragment>
`;
